### ISSUE TITLE
How can I customize the profile.bin ？

### ISSUE THREAD

**Issue by jricoclq** [zoffline/zwift-offline#272]
How can I customize the profile.bin ？ I don't want to use a level 50 attribute profile！

**Comment by fatsbrown** [comment]
You can customize it in the "initials" icon (my profile) in the home screen (next to the exit icon) or in the "pencil" icon under your name in the pause menu.

You don't need to use a level 50 profile, start without a profile.bin and a new one will be created.

**Comment by jricoclq** [comment]
i meam how can i customize the file "profile.bin" .

**Comment by fatsbrown** [comment]
Yes, the file profile.bin stores your profile data, which can be customized in the place I told you. Other than that, it stores your progress. What exactly do you want to customize?

**Comment by fatsbrown** [comment]
If you wish to customize the avatar, it's in the "garage" icon.

https://github.com/zoffline/zwift-offline/assets/123950591/32e68956-a6c3-47bd-918a-d98611c3e649

**Comment by fatsbrown** [comment]
profile.bin is a serialized PlayerProfile protobuf message
https://github.com/zoffline/zwift-offline/blob/69a4961c8a1b6c80523c1d64813ba92726724b93/protobuf/profile.proto#L33

If you want to change fields that are not possible to change in Zwift interface you can do this (run from zoffline source directory):

Read and parse the file
```python
import protobuf.profile_pb2
profile = protobuf.profile_pb2.PlayerProfile()
profile_file = 'storage/1/profile.bin'
with open(profile_file, 'rb') as f:
    profile.ParseFromString(f.read())
```
Change what you want, for example
```python
profile.total_xp = 500000
```
Serialize and write back to file
```python
with open(profile_file, 'wb') as f:
    f.write(profile.SerializeToString())
```

**Comment by jricoclq** [comment]
> profile.bin is a serialized PlayerProfile protobuf message
> 
> https://github.com/zoffline/zwift-offline/blob/69a4961c8a1b6c80523c1d64813ba92726724b93/protobuf/profile.proto#L33
> 
> If you want to change fields that are not possible to change in Zwift interface you can do this (run from zoffline source directory):
> 
> Read and parse the file
> 
> ```python
> import protobuf.profile_pb2
> profile = protobuf.profile_pb2.PlayerProfile()
> profile_file = 'storage/1/profile.bin'
> with open(profile_file, 'rb') as f:
>     profile.ParseFromString(f.read())
> ```
> 
> Change what you want, for example
> 
> ```python
> profile.total_xp = 500000
> ```
> 
> Serialize and write back to file
> 
> ```python
> with open(profile_file, 'wb') as f:
>     f.write(profile.SerializeToString())
> ```

OK,TKS a lot!

**Comment by jricoclq** [comment]
> profile.bin is a serialized PlayerProfile protobuf message
> 
> [zwift-offline/protobuf/profile.proto](https://github.com/zoffline/zwift-offline/blob/69a4961c8a1b6c80523c1d64813ba92726724b93/protobuf/profile.proto#L33)
> 
> Line 33 in [69a4961](/zoffline/zwift-offline/commit/69a4961c8a1b6c80523c1d64813ba92726724b93)
> 
>  message PlayerProfile { 
> If you want to change fields that are not possible to change in Zwift interface you can do this (run from zoffline source directory):
> 
> Read and parse the file
> 
> import protobuf.profile_pb2
> profile = protobuf.profile_pb2.PlayerProfile()
> profile_file = 'storage/1/profile.bin'
> with open(profile_file, 'rb') as f:
>     profile.ParseFromString(f.read())
> Change what you want, for example
> 
> profile.total_xp = 500000
> Serialize and write back to file
> 
> with open(profile_file, 'wb') as f:
>     f.write(profile.SerializeToString())

May I ask if it is possible to complete the new challenge in this way? Can you tell me which fields can be used to modify this?

**Comment by fatsbrown** [comment]
> May I ask if it is possible to complete the new challenge in this way? Can you tell me which fields can be used to modify this?

It's stored in field 155, you need to add it to `PlayerProfile` in profile.proto and compile with `protoc --python_out=. profile.proto`
```
	optional float factory_tour_challenge = 155;
```

![Image](https://github.com/user-attachments/assets/f6a5e532-f0f1-46bb-ac9f-37a2e034aeaf)

**Comment by jricoclq** [comment]
> > May I ask if it is possible to complete the new challenge in this way? Can you tell me which fields can be used to modify this?
> 
> It's stored in field 155, you need to add it to `PlayerProfile` in profile.proto and compile with `protoc --python_out=. profile.proto`
> 
> ```
> 	optional float factory_tour_challenge = 155;
> ```
> 
> ![Image](https://github.com/user-attachments/assets/f6a5e532-f0f1-46bb-ac9f-37a2e034aeaf)

Am I doing this right?
![Image](https://github.com/user-attachments/assets/3055bec0-9e0d-48f6-bb71-ee86da9a3338)

**Comment by fatsbrown** [comment]
You should add in the end.

![Image](https://github.com/user-attachments/assets/7800b63c-9e86-4834-a434-77d63d033052)

**Comment by jricoclq** [comment]
> 你应该在最后添加。
> 
> ![Image](https://github.com/user-attachments/assets/7800b63c-9e86-4834-a434-77d63d033052)

How did this error occur? I have tried many times and have been told that this field is not available?
![Image](https://github.com/user-attachments/assets/1e68a98c-508d-416b-a9c3-1388c7e4d764)

**Comment by jricoclq** [comment]
Can you please send me the profile. proto file that you have compiled so that I can try it out? I think I did not edit it correctly.

**Comment by fatsbrown** [comment]
Download from https://github.com/oldnapalm/zwift-offline/blob/fitness/protobuf/profile_pb2.py

Notice that in this file the field is named `power_units`

https://github.com/oldnapalm/zwift-offline/blob/7cba431c39796e4c68d2945cf2559828536fdf5c/protobuf/profile.proto#L223

**Comment by jricoclq** [comment]
> Download from https://github.com/oldnapalm/zwift-offline/blob/fitness/protobuf/profile_pb2.py
> 
> Notice that in this file the field is named `power_units`
> 
> https://github.com/oldnapalm/zwift-offline/blob/7cba431c39796e4c68d2945cf2559828536fdf5c/protobuf/profile.proto#L223

At first, I thought that wearing lightning socks would result in a special effect, but it didn't appear. I thought it was because the challenge wasn't completed, so it didn't. Now that the challenge has been completed, the special effect still hasn't appeared. It seems that only the Fire Socks have special effects locally.

![Image](https://github.com/user-attachments/assets/707b7acb-df9d-4898-a089-ebe21c0cc4a9)

![Image](https://github.com/user-attachments/assets/c8dab73b-793d-4240-98b4-b075d625b182)

**Comment by jricoclq** [comment]
> Download from https://github.com/oldnapalm/zwift-offline/blob/fitness/protobuf/profile_pb2.py
> 
> Notice that in this file the field is named `power_units`
> 
> https://github.com/oldnapalm/zwift-offline/blob/7cba431c39796e4c68d2945cf2559828536fdf5c/protobuf/profile.proto#L223

Thank you for your help , tks very musch!

**Comment by fatsbrown** [comment]
Maybe the visual effect screenshot is from a development version. Even the field `bUseSockCDA` of `bikemodifiercalculator_params` in variants.txt is still disabled, it's impossible that someone already unlocked the socks in just a few days.

**Comment by jricoclq** [comment]
> Maybe the visual effect screenshot is from a development version. Even the field `bUseSockCDA` of `bikemodifiercalculator_params` in variants.txt is still disabled, it's impossible that someone already unlocked the socks in just a few days.

You're right, it's like the "Bike Upgrades & Garage Sorting" hasn't been launched yet. Originally, it was supposed to be 2024Q4. Hahaha
